bokeh/io.py

```diff
@@ -489,25 +489,79 @@ def save(obj, filename=None, resources=None, title=None, state=None, **kwargs):
     _save_helper(obj, filename, resources, title)
     return os.path.abspath(filename)
 
-def _detect_filename(ext):
-    """ Detect filename from the name of the script being run. Returns
-    temporary file if the script could not be found or the location of the
-    script does not have write permission (e.g. interactive mode).
-    """
+def _no_access(basedir):
+    ''' Return True if the given base dir is not accessible or writeable
+
+    '''
+    return not os.access(basedir, os.W_OK | os.X_OK)
+
+def _shares_exec_prefix(basedir):
+    ''' Whether a give base directory is on the system exex prefix
+
+    '''
+    import sys
+    prefix = sys.exec_prefix
+    return (prefix is not None and basedir.startswith(prefix))
+
+def _temp_filename(ext):
+    ''' Generate a temporary, writable filename with the given extension
+
+    '''
+    return tempfile.NamedTemporaryFile(suffix="." + ext).name
+
+def _detect_current_filename():
+    ''' Attempt to return the filename of the currently running Python process
+
+    Returns None if the filename cannot be detected.
+    '''
     import inspect
-    from os.path import dirname, basename, splitext, join, curdir
 
+    filename = None
     frame = inspect.currentframe()
-    while frame.f_back and frame.f_globals.get('name') != '__main__':
-        frame = frame.f_back
+    try:
+        while frame.f_back and frame.f_globals.get('name') != '__main__':
+            frame = frame.f_back
+
+        filename = frame.f_globals.get('__file__')
+    finally:
+        del frame
+        return filename
```

Comment:
This isn't a big issue, but it would recommend not using `return` statement in `finally` block. This is because the body `finally` block doesn't participate normal execution path, so it may be confusing when someone adds (perhaps accidentally) another `return` statement.